### PR TITLE
Use NFI for String#crypt

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 1]
```diff
@@ -788,62 +788,6 @@ private int countSlow(DynamicObject string, DynamicObject... otherStrings) {
...
-
-            final DynamicObject ret = createString(RopeOperations.ropeFromByteList(RopeBuilder.createRopeBuilder(cryptedString, 0, cryptedString.length - 1, ASCIIEncoding.INSTANCE)));
-
-            taintResultNode.maybeTaint(string, ret);
-            taintResultNode.maybeTaint(salt, ret);
-
-            return ret;
-        }
-
-    }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: That's a big one to start with!

File: src/main/ruby/core/posix.rb [hunk 2]
```diff
@@ -63,7 +63,9 @@ def self.attach_function(method_name, native_name = method_name, argument_types,
           args[i] = str
         end
 
-        bound_func.call(*args)
+        result = bound_func.call(*args)
+        result = Truffle::Interop.from_java_string(Truffle::Interop.unbox(result)) if return_type == :string
```
Comment (eregon) [inline src/main/ruby/core/posix.rb]: Using UNBOX here will assume the bytes have a UTF-8 encoding, which might be wrong for instance for `read()`/`write()`. All sequences of bytes are also not all valid UTF-8 strings, so then it's just impossible to encode in a j.l.String, unless both side are aware of it and use the char[] as a byte[] (but how can they communicate that if they also have to handle textual UTF-8 bytes?).
Additionally, this does 2 re-encodings of the String, so reading bytes directly is better.

We can automate the conversion if the `return_type` is String, it just has to use the logic below (or similar) and handle a returned NULL pointer.

Comment (chrisseaton) [reply src/main/ruby/core/posix.rb]: How's this do you think? (`.nil?` on NFI's objects does work.)

```ruby
        if return_type == :string
          if result.nil?
            nil
          else
            ptr = Rubinius::FFI::Pointer.new(Truffle::Interop.as_pointer(result))
            result = ptr.read_string_to_null
          end
        end
```

Comment (eregon) [reply src/main/ruby/core/posix.rb]: Yes, that's good, that's what I had below in `getenv`.

File: src/main/ruby/core/posix.rb [hunk 3]
```diff
@@ -132,8 +143,7 @@ def self.getenv(name)
     if value.nil?
       nil
     else
-      ptr = Rubinius::FFI::Pointer.new(Truffle::Interop.as_pointer(value))
-      ptr.read_string_to_null
+      value
     end
   end
```
Comment (eregon) [inline src/main/ruby/core/posix.rb]: The entire `getenv` method can be removed now that a return_type of :string does the right thing, and then the attach_function is enough.

File: src/main/ruby/core/posix.rb (hunk 2 above)
Comment (eregon) [reply src/main/ruby/core/posix.rb]: And yes it's quite nice that .nil? sends IS_NULL :)

**Pull request merged**